drivers/clock_control/clock_control_it8xxx2.c

```diff
@@ -0,0 +1,337 @@
+	/* Pre-set FND clock frequency = PLL / 3 */
+	ecpm_regs->SCDCR0 = (2 << 4);
+	/* JTAG and EC */
+	ecpm_regs->SCDCR3 = (pll->div_jtag << 4) | pll->div_ec;
+	/* Chip sleep after wait for interrupt (wfi) instruction */
+	chip_pll_ctrl(CHIP_PLL_SLEEP);
+	/* Chip sleep and wait timer wake it up */
+	__asm__ volatile  ("wfi");
+	/* New FND clock frequency */
+	ecpm_regs->SCDCR0 = pll->div_fnd << 4;
+	/* Chip doze after wfi instruction */
+	chip_pll_ctrl(CHIP_PLL_DOZE);
+	/* UART */
+	ecpm_regs->SCDCR1 = pll->div_uart;
+	/* SSPI and SMB */
+	ecpm_regs->SCDCR2 = (pll->div_sspi << 4) | pll->div_smb;
+	/* USBPD and PWM */
+	ecpm_regs->SCDCR4 = (pll->div_usbpd << 4) | pll->div_pwm;
+}
+
+static void chip_configure_pll(const struct pll_config_t *pll)
+{
+	struct ecpm_it8xxx2_regs *const ecpm_regs = clock_config->reg_base;
+
+	/* Re-configure PLL clock or not. */
+	if (((ecpm_regs->PLLFREQR & PLLFREQ_MASK) != pll->pll_setting) ||
+	    ((ecpm_regs->SCDCR0 & FNDFREQ_MASK) != (pll->div_fnd << 4)) ||
+	    ((ecpm_regs->SCDCR3 & EC_FREQ_MASK) != pll->div_ec)) {
+#ifdef CONFIG_ESPI
+		/*
+		 * We have to disable eSPI pad before changing
+		 * PLL sequence or sequence will fail if CS# pin is low.
+		 */
+		espi_it8xxx2_enable_pad_ctrl(ESPI_IT8XXX2_SOC_DEV, false);
+#endif
+		/* Run change PLL sequence */
+		chip_run_pll_sequence(pll);
+#ifdef CONFIG_ESPI
+		/* Enable eSPI pad after changing PLL sequence */
+		espi_it8xxx2_enable_pad_ctrl(ESPI_IT8XXX2_SOC_DEV, true);
+#endif
+	}
+}
+
+#endif /* CONFIG_SOC_IT8XXX2_PLL_FLASH_48M */
+
+static int it8xxx2_clock_control_init(const struct device *dev)
+{
+	clock_config = dev->config;
+	int pllfreq = clock_config->pll_freq;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = clock_config->reg_base;
+
+	/* SWUC Clock Gating */
+	ecpm_regs->CGCTRL2R |= SWUC_CLOCK_GATING;
+	/* PECI, SSPI, DBGR Clock Gating */
+	ecpm_regs->CGCTRL3R |= PECI_SSPI_DBGR_CLOCK_GATING;
+	/* SMB/I2C Clock Gating */
+	ecpm_regs->CGCTRL4R |= SMB_ALL_CHANNELS_CLOCK_GATING;
+
+
+	if (!(DT_NODE_HAS_STATUS(DT_NODELABEL(usbpd0), okay))) {
+		/* PD0 Clock Gating */
+		ecpm_regs->CGCTRL5R |= PD0_CLOCK_GATING;
+	}
+
+	if (!(DT_NODE_HAS_STATUS(DT_NODELABEL(usbpd1), okay))) {
+		/* PD1 Clock Gating */
+		ecpm_regs->CGCTRL5R |= PD1_CLOCK_GATING;
+	}
+
+#ifdef CONFIG_ESPI_IT8XXX2
+	/* SPI SLAVE Clock Gating */
+	ecpm_regs->CGCTRL5R |= SPI_SLAVE_CLOCK_GATING;
+#endif /* #ifdef CONFIG_ESPI_IT8XXX2 */
+
+#ifndef CONFIG_SOC_IT8XXX2_JTAG_ENABLE
+	/* JTAG Clock Gating */
+	ecpm_regs->CGCTRL6R |= JTAG_CHB_CLOCK_GATING;
+#endif /* #ifdef CONFIG_SOC_IT8XXX2_JTAG_ENABLE */
+
+#ifndef CONFIG_SOC_IT8XXX2_ENABLE_FPU
+	/* FPU Clock Gating */
+	ecpm_regs->CGCTRL6R |= FPU_CHA_CLOCK_GATING;
+#endif
+
+	/* PLL Config */
+	if (pllfreq == PLL_DEFAULT) {
+		return 0;
+	}
+
+	if (IS_ENABLED(CONFIG_ITE_IT8XXX2_INTC)) {
+		ite_intc_save_and_disable_interrupts();
+	}
+	/* configure PLL/CPU/flash clock */
+	chip_configure_pll(&pll_configuration[pllfreq]);
```

Comment:
You could move the PLL frequency set into `it8xxx2_clock_control_set_rate()`.  This call would then be `it8xxx2_clock_control_set_rate(dev, IT8XXX2_PLL, pllfreq);`

This would provide flexibility to the application to change the PLL and the CPU frequency.

Reply:
@keith-zephyr 

As mentioned in previous conversation, we'd love to add the flexibility into the code but even that the code will not open the PLL change in the real-time for design reasons. However, the flexibility should be somehow contained in the current update.

If there's any further idea/suggestion, please kindly let me know.
Thanks again for patience & guidance.